src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,695 @@
+
+        /// <summary>
+        /// Creates a priority queue populated with the specified elements and priorities.
+        /// </summary>
+        public PriorityQueue(IEnumerable<(TElement element, TPriority priority)> items)
+            : this(items, comparer: null)
+        {
+        }
+
+        /// <summary>
+        /// Creates a priority queue populated with the specified elements and priorities,
+        /// and with the specified priority comparer.
+        /// </summary>
+        public PriorityQueue(IEnumerable<(TElement element, TPriority priority)> items, IComparer<TPriority>? comparer)
+        {
+            if (items is null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            _nodes = EnumerableHelpers.ToArray(items, out _size);
+            _unorderedItems = new Lazy<UnorderedItemsCollection>(() => new UnorderedItemsCollection(this));
+            this.Comparer = comparer ?? Comparer<TPriority>.Default;
+
+            if (_size > 1)
+            {
+                Heapify();
+            }
+        }
+
+        /// <summary>
+        /// Gets the current amount of items in the priority queue.
+        /// </summary>
+        public int Count => _size;
+
+        /// <summary>
+        /// Gets the priority comparer of the priority queue.
+        /// </summary>
+        public IComparer<TPriority> Comparer { get; }
+
+        /// <summary>
+        /// Gets a collection that enumerates the elements of the queue.
+        /// </summary>
+        public UnorderedItemsCollection UnorderedItems => _unorderedItems.Value;
+
+        /// <summary>
+        /// Enqueues the specified element and associates it with the specified priority.
+        /// </summary>
+        public void Enqueue(TElement element, TPriority priority)
+        {
+            if (element is null)
+            {
+                throw new ArgumentNullException(nameof(element));
+            }
+
+            if (priority is null)
+            {
+                throw new ArgumentNullException(nameof(priority));
+            }
+
+            EnsureEnoughCapacityBeforeAddingNode();
+
+            // Add the node at the end
+            var node = (element, priority);
+            _nodes[_size] = node;
+            _size++;
+            _version++;
+
+            // Restore the heap order
+            int lastNodeIndex = GetLastNodeIndex();
+            MoveUp(node, lastNodeIndex);
+        }
+
+        /// <summary>
+        /// Gets the element associated with the minimal priority.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The queue is empty.</exception>
+        public TElement Peek()
+        {
+            if (TryPeek(out TElement? element, out TPriority? priority))
+            {
+                return element;
+            }
+            else
+            {
+                throw new InvalidOperationException(SR.InvalidOperation_EmptyQueue);
+            }
```

Comment:
Nit: the else is unnecessary, as is naming the priority (as it's never used).  This can either be:
```C#
if (!TryPeek(out TElement? element, out _))
{
    throw new InvalidOperationException(SR.InvalidOperation_EmptyQueue);
}

return element;
```
or
```C#
if (TryPeek(out TElement? element, out _))
{
    return element;
}

throw new InvalidOperationException(SR.InvalidOperation_EmptyQueue);
```
or even if you want real brevity:
```C#
public TElement Peek() =>
    TryPeek(out TElement? element, out _) ?
        element :
        throw new InvalidOperationException(SR.InvalidOperation_EmptyQueue);
```